### PR TITLE
Add flang-activation

### PR DESCRIPTION
**Pull request by h-vetinari**
Address https://github.com/conda-forge/flang-feedstock/issues/63

### CODE REVIEW

**Comment by h-vetinari** [comment]
@isuruf, this should handle the flang compiler activation split, PTAL! :)

I added you as a maintainer, please let me know if you'd prefer not to be.

**Comment by h-vetinari** [comment]
Gentle ping @isuruf

File: recipes/flang-activation/activate.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+export _OLD_FC=$FC
+export _OLD_CPATH=$CPATH
+export _OLD_FFLAGS=$FFLAGS
+export _OLD_LIBRARY_PATH=$LIBRARY_PATH
+export _OLD_FORTRANFLAGS=$FORTRANFLAGS
+
+if [ -z "$CONDA_PREFIX" ]; then
+    export CPATH=$CPATH:$PREFIX/include
+    export LIBRARY_PATH=$LIBRARY_PATH:$PREFIX/lib
```
Comment (isuruf) [inline recipes/flang-activation/activate.sh:9]: Let's not set these. We should just set the correct flags in `FFLAGS`

File: recipes/flang-activation/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    test:
+      requires:
+        - grep                      # [unix]
+        - m2-grep                   # [win]
+      files:
+        - hello_world.f90
+      commands:
+        # see naming discussion: https://discourse.llvm.org/t/reviving-rename-flang-new-to-flang/68130/2
+        # - flang hello_world.f90
+        - flang-new hello_world.f90
```
Comment (isuruf) [inline recipes/flang-activation/meta.yaml:48]:
```suggestion
        - $FC hello_world.f90
```

File: recipes/flang-activation/activate.sh (hunk 1 above)
Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:9]: Happy to! For context, is untouched from the existing [activation](https://github.com/conda-forge/flang-feedstock/blame/main/recipe/activate.sh) in flang.

**Comment by h-vetinari** [comment]
Thanks for the review - that should all be fixed now.

File: recipes/flang-activation/activate.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export _OLD_FC=$FC
+export _OLD_FFLAGS=$FFLAGS
+export _OLD_FORTRANFLAGS=$FORTRANFLAGS
```
Comment (isuruf) [inline recipes/flang-activation/activate.sh:3]: Let's use `_CONDA_BACKUP` instead of `_OLD` to match other compilers.

File: recipes/flang-activation/activate.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+
+export FORTRANFLAGS="$FORTRANFLAGS --sysroot=$CONDA_BUILD_SYSROOT"
+
+export FC="flang-new"
+export FFLAGS=$FORTRANFLAGS
```
Comment (isuruf) [inline recipes/flang-activation/activate.sh:8]:
```suggestion
export FORTRANFLAGS="$FORTRANFLAGS"

export FC="flang-new"
export FFLAGS="$FFLAGS  --sysroot=$CONDA_BUILD_SYSROOT"
```

Comment (isuruf) [reply recipes/flang-activation/activate.sh:8]: With `gcc_linux-64`, CONDA_BUILD_SYSROOT is not set right?

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:8]: I don't really know the difference between `FFLAGS` and `FORTRANFLAGS`; we added the sysroot to both in https://github.com/conda-forge/flang-feedstock/commit/267195daf179af87ec0812b9c4c1164337b6aae9, which was when we [switched](https://github.com/conda-forge/flang-feedstock/pull/28) to llvm-flang.

To answer your question, AFAICT `gcc_linux-64` [does set](https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/e22827cd872402d807ae3452e23673d6d3e4b0b0/recipe/activate-gcc.sh#L187) `CONDA_BUILD_SYSROOT`

Comment (isuruf) [reply recipes/flang-activation/activate.sh:8]: Sorry, I meant without `gcc_linux-64` in run, we won't have `CONDA_BUILD_SYSROOT`, right?

File: recipes/flang-activation/activate.sh (hunk 3 above)
Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:3]: Fine by me. GCC [uses](https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/e22827cd872402d807ae3452e23673d6d3e4b0b0/recipe/activate-gcc.sh#L45) `CONDA_BACKUP` without a leading underscore, so I'll follow that for consistency.

File: recipes/flang-activation/activate.sh (hunk 4 above)
Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:8]: Yeah, it seems we're only setting (rather than using) that in `gcc_{{ target }}` and [`clang_{{ target }}`](https://github.com/conda-forge/clang-compiler-activation-feedstock/blob/52e3513c4592a367930372717c22cc5fb680d133/recipe/activate-clang.sh#L197).

Do you want us to remove the `--sysroot` flag here, or reconstruct `CONDA_BUILD_SYSROOT`?

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:8]: or depend on `gcc_linux64`...

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:8]: > Do you want us to remove the `--sysroot` flag here

At least the helloworld example on the flang feedstock passes without that it seems (no `gcc_linux-64` in environment, `CONDA_BUILD_SYSROOT` not set):

<details>

```
TEST START: /home/conda/feedstock_root/build_artifacts/linux-64/flang-19.1.0.rc3-h43464a6_0.conda
[...]
The following NEW packages will be INSTALLED:

    _libgcc_mutex:                        0.1-conda_forge               conda-forge              
    _openmp_mutex:                        4.5-2_gnu                     conda-forge              
    _sysroot_linux-64_curr_repodata_hack: 3-h69a702a_16                 conda-forge              
    binutils_impl_linux-64:               2.40-ha1999f0_7               conda-forge              
    bzip2:                                1.0.8-h4bc722e_7              conda-forge              
    clang:                                19.1.0.rc3-default_he6a287e_0 conda-forge/label/llvm_rc
    clang-19:                             19.1.0.rc3-default_hc83ec93_0 conda-forge/label/llvm_rc
    flang:                                19.1.0.rc3-h43464a6_0         local                    
    grep:                                 3.11-h7af0fdc_3               conda-forge              
    icu:                                  75.1-he02047a_0               conda-forge              
    kernel-headers_linux-64:              3.10.0-h4a8ded7_16            conda-forge              
    ld_impl_linux-64:                     2.40-hf3520f5_7               conda-forge              
    libclang-cpp19.1:                     19.1.0.rc3-default_hc83ec93_0 conda-forge/label/llvm_rc
    libflang:                             19.1.0.rc3-h43464a6_0         local                    
    libgcc-devel_linux-64:                14.1.0-h5d3d1c9_100           conda-forge
    [...]

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
export PREFIX=/home/conda/feedstock_root/build_artifacts/flang-split_1724314217473/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac
export SRC_DIR=/home/conda/feedstock_root/build_artifacts/flang-split_1724314217473/test_tmp
+ flang-new --sysroot= hello_world.f90
+ ./a.out
 Hello World!
```

</details>

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh]: I've removed that flag for now. If we do need it, we should IMO add a test in the activation-feedstock that exercise the necessity to use the sysroot.

**Comment by h-vetinari** [comment]
Anything else here @isuruf?

**Comment by h-vetinari** [comment]
Gentle ping @isuruf

File: recipes/flang-activation/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+  # intentionally only windows (main target) & linux (debuggability)
+  skip: true  # [osx]
+
+outputs:
+  - name: flang_impl_{{ cross_target_platform }}
+    script: install_flang_impl.sh   # [unix]
+    # we don't cross-compile on windows -> no symlinks
+    requirements:
+      - flang ={{ version }}
+      - lld ={{ version }}          # [win]
```
Comment (isuruf) [inline recipes/flang-activation/meta.yaml:18]: Do we need the exact version here?

File: recipes/flang-activation/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      - compiler-rt ={{ version }}
```
Comment (isuruf) [inline recipes/flang-activation/meta.yaml:19]: Is the package without `_{{ cross_target_platform }}` needed?

File: recipes/flang-activation/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        - libflang >={{ version }}
+    requirements:
+      run:
+        - {{ pin_subpackage("flang_impl_" ~ cross_target_platform, exact=True) }}
+    test:
+      files:
+        - hello_world.f90
+      commands:
+        # rename flang-new -> flang is happening soon:
+        # https://discourse.llvm.org/t/proposal-rename-flang-new-to-flang/69462
```
Comment (isuruf) [inline recipes/flang-activation/meta.yaml:42]: This comment seems irrelevant

File: recipes/flang-activation/meta.yaml (hunk 5 above)
Comment (h-vetinari) [reply recipes/flang-activation/meta.yaml:18]: Should be fine with any recent lld I guess.

File: recipes/flang-activation/meta.yaml (hunk 6 above)
Comment (h-vetinari) [reply recipes/flang-activation/meta.yaml:19]: I was basing this on [`clang_linux-64`](https://github.com/conda-forge/ctng-compiler-activation-feedstock/blob/main/recipe/meta.yaml#L97), where this is specified like this. Happy to remove

File: recipes/flang-activation/meta.yaml (hunk 7 above)
Comment (h-vetinari) [reply recipes/flang-activation/meta.yaml:42]: This is an updated version from what we currently [have](https://github.com/conda-forge/flang-feedstock/blob/35b55990fe7350ff4dcde7a3a1298714ca6f5660/recipe/meta.yaml#L121-L122) in the flang feedstock. As it was looking like it would finally happen a few days ago (until the goalposts shifted yet again), I was debating with myself whether we want to do the rename for flang 19 already. Might make the fortran migration a bit simpler. 🤷

Comment (isuruf) [reply recipes/flang-activation/meta.yaml:42]: I didn't mean the contents of the comment. I was referring to why this comment needs to be in the test section. It seems irrelevant.

File: recipes/flang-activation/conda_build_config.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+cross_target_platform:
+  - linux-64                    # [linux]
+# - emscripten-wasm32           # [linux]
+  - win-64                      # [win]
+  - dummy                       # [osx]
+
+zip_keys:
+  -
+    - CBUILD
```
Comment (isuruf) [inline recipes/flang-activation/conda_build_config.yaml:21]: This shouldn't be in the zip_keys with the others. Not necessary for now since there's no cross-compilation, but will be useful in the future,

File: recipes/flang-activation/activate.sh [hunk 9]
```diff
@@ -0,0 +1,2 @@
+export CONDA_BACKUP_FC=$FC
+export FC="flang-new"
```
Comment (isuruf) [inline recipes/flang-activation/activate.sh:2]:
```suggestion
export FC="@CHOST@-flang"
```

Comment (isuruf) [reply recipes/flang-activation/activate.sh:2]: with @CHOST@ replaced

File: recipes/flang-activation/meta.yaml (hunk 7 above)
Comment (h-vetinari) [reply recipes/flang-activation/meta.yaml]: It's more visible in the recipe than in the activation script, and the test section is where we're using the name (or now, the value of `FC`).

File: recipes/flang-activation/activate.bat [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+set "CONDA_BACKUP_LD=%LD%"
+set "CONDA_BACKUP_LDFLAGS=%LDFLAGS%"
+
+:: flang 19 still uses "temporary" name
+set "FC=flang-new"
+set "LD=lld-link.exe"
+
+:: following https://github.com/conda-forge/clang-win-activation-feedstock/blob/main/recipe/activate-clang_win-64.bat
+set "FFLAGS=-D_CRT_SECURE_NO_WARNINGS -fms-runtime-lib=dll -fuse-ld=lld -I%LIBRARY_INC%"
+set "LDFLAGS=-Wl,-defaultlib:%CONDA_PREFIX:\=/%/lib/clang/@MAJOR_VER@/lib/windows/clang_rt.builtins-x86_64.lib"
```
[the remark this answers is not in the document]
Comment (h-vetinari) [reply recipes/flang-activation/activate.bat:12]: But we need to link to compiler-rt, otherwise things don't work.

File: recipes/flang-activation/activate.sh (hunk 9 above)
Comment (h-vetinari) [reply recipes/flang-activation/activate.sh:2]: Unix-only? Or also on windows?

File: recipes/flang-activation/activate.bat (hunk 10 above)
[the remark this answers is not in the document]
Comment (isuruf) [reply recipes/flang-activation/activate.bat:12]: Yeah, let's make it additive, otherwise you'll override the values from `clang_win-64` if both are used in a recipe.

File: recipes/flang-activation/activate.sh (hunk 9 above)
Comment (isuruf) [reply recipes/flang-activation/activate.sh:2]: Unix-only. Windows is annoying due to no symlinks.

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh]: On windows this isn't going to work so easily, since we don't have the symlinks.

Comment (h-vetinari) [reply recipes/flang-activation/activate.sh]: sorry, my tab didn't refresh to show your comments

File: recipes/flang-activation/conda_build_config.yaml [hunk 11]
```diff
@@ -0,0 +1,28 @@
+CBUILD:
+  - x86_64-conda-linux-gnu      # [linux]
+# - x86_64-conda-linux-gnu      # [linux]
```
Comment (isuruf) [inline recipes/flang-activation/conda_build_config.yaml:3]:
```suggestion
```

**Review by isuruf — APPROVED** [review]

**Comment by h-vetinari** [comment]
Thanks for iterating on this. We can keep polishing things on the feedstock, flang 19.1.0 GA is around the corner anyway. :)
